p4-16/spec/grammar.adoc

```diff
@@ -852,74 +858,65 @@ switchLabel
     ;
 // end::switchLabel[]
 
-// tag::loopStatement[]
-loopStatement
-    : optAnnotations FOR "(" optForInitStatements ";" expression ";"
-      optForUpdateStatements ")" statement
-    | optAnnotations FOR "(" typeRef name IN forCollectionExpr ")" statement
+// tag::statementOrDeclaration[]
+statementOrDeclaration
+    : variableDeclaration
+    | constantDeclaration
+    | statement
     ;
-// end::loopStatement[]
+// end::statementOrDeclaration[]
 
-// tag::optForInitStatements[]
-optForInitStatements
-    : /* empty */
-    | forInitStatements
+// tag::forStatement[]
+forStatement
+    : optAnnotations FOR "(" forInitStatements ";" expression ";"
+      forUpdateStatements ")" statement
+    | optAnnotations FOR "(" typeRef name IN forCollectionExpr ")" statement
+    | optAnnotations FOR "(" annotations typeRef name IN forCollectionExpr ")" statement
     ;
-// end::optForInitStatements[]
+// end::forStatement[]
 
 // tag::forInitStatements[]
 forInitStatements
-    : forInitStatement
-    | forInitStatements "," forInitStatement
+    : /* empty */
+    | forInitStatementsNonEmpty
     ;
 // end::forInitStatements[]
 
-// tag::forInitStatement[]
-forInitStatement
-    : optAnnotations typeRef name optInitializer
-    | lvalue "(" argumentList ")"
-    | lvalue "<" typeArgumentList ">" "(" argumentList ")"
-    | lvalue "=" expression
+// tag::forInitStatementsNonEmpty[]
+forInitStatementsNonEmpty
+    : declOrAssignmentOrMethodCallStatement
```

Comment:
I'm guessing we currently allow all of these in the P4C grammar? But I wonder if we should instead be more restrictive and e.g. only allow a variable declaration? Or perhaps a variable declaration or normal assignment?

Reply:
We certainly could make such restrictions.  It would of course limit the kinds of for loops a developer can write, accordingly.

We could also restrict things even further, e.g. no init statements and no update statements, and it would still be possible to express all possible loops that the current grammar allows.  Thus this isn't as much a question about "what is possible to express", as it is "how easily do we wish to allow developers to express certain things, without working around such restrictions?".

Reply:
Right. With the trade-off being grammar complexity vs expressiveness.

This approach seems OK to me, and I buy your point below about textual similarity being pretty useful, so resolving for now.